core/src/main/java/io/grpc/internal/DelayedStream.java

```diff
@@ -86,182 +95,224 @@ public PendingMessage(InputStream message, boolean shouldBeCompressed) {
-   * {@code cancelledPrematurely} and this method atomically.
-   */
-  void setStream(ClientStream stream) {
+  @VisibleForTesting
+  void startStream() {
     synchronized (this) {
-      if (cancelledPrematurely()) {
-        // Already cancelled
-        throw new IllegalStateException("Can't set on cancelled stream");
-      }
-      checkState(realStream == null, "Stream already created: %s", realStream);
-      realStream = stream;
+      checkState(unstartedStream != null, "unstartedStream");
+      checkState(listener != null, "listener");
+      unstartedStream.start(listener);
+
       if (compressionMessageEncodings != null) {
-        realStream.pickCompressor(compressionMessageEncodings);
+        unstartedStream.pickCompressor(compressionMessageEncodings);
       }
       if (this.decompressionRegistry != null) {
-        realStream.setDecompressionRegistry(this.decompressionRegistry);
+        unstartedStream.setDecompressionRegistry(this.decompressionRegistry);
       }
       if (this.compressionRegistry != null) {
-        realStream.setCompressionRegistry(this.compressionRegistry);
+        unstartedStream.setCompressionRegistry(this.compressionRegistry);
       }
       for (PendingMessage message : pendingMessages) {
-        realStream.setMessageCompression(message.shouldBeCompressed);
-        realStream.writeMessage(message.message);
+        unstartedStream.setMessageCompression(message.shouldBeCompressed);
+        unstartedStream.writeMessage(message.message);
       }
       // Set this again, incase no messages were sent.
-      realStream.setMessageCompression(messageCompressionEnabled);
+      unstartedStream.setMessageCompression(messageCompressionEnabled);
       pendingMessages.clear();
       if (pendingHalfClose) {
-        realStream.halfClose();
+        unstartedStream.halfClose();
         pendingHalfClose = false;
       }
       if (pendingFlowControlRequests > 0) {
-        realStream.request(pendingFlowControlRequests);
+        unstartedStream.request(pendingFlowControlRequests);
         pendingFlowControlRequests = 0;
       }
       if (pendingFlush) {
-        realStream.flush();
+        unstartedStream.flush();
         pendingFlush = false;
       }
+      // Ensures visibility.
+      startedStream = unstartedStream;
     }
   }
 
-  void maybeClosePrematurely(final Status reason) {
+  void setStream(ClientStream stream) {
     synchronized (this) {
-      if (realStream == null) {
-        realStream = NoopClientStream.INSTANCE;
-        listener.closed(reason, new Metadata());
+      if (error != null) {
+        // If there is an error, realStream will be a Noop.
+        return;
+      }
+      checkState(unstartedStream == null, "Stream already created: %s", unstartedStream);
+      unstartedStream = checkNotNull(stream, "stream");
+      // listener can only be non-null if start has already been called.
+      if (listener != null) {
```

Comment:
Doing so makes it so that stream.start would not be the last call in CCI.start.  I like having it be the last call, because it makes it visually clear that the stream is safe to keep having setup calls on it without fear of listener being invoked (or other network traffic).